geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java

```diff
@@ -127,50 +127,139 @@ private int getArrayIndex(int listIndex) {
   }
 
   /**
-   * @param elementsToAdd elements to add to this set; NOTE this list may by modified by this call
-   * @param region the region this instance is stored in
-   * @param key the name of the set to add to
-   * @param onlyIfExists if true then the elements should only be added if the key already exists
-   *        and holds a list, otherwise no operation is performed.
-   * @return the length of the list after the operation
-   */
-  public long lpush(List<byte[]> elementsToAdd, Region<RedisKey, RedisData> region, RedisKey key,
-      final boolean onlyIfExists) {
-    elementsPush(elementsToAdd);
-    storeChanges(region, key, new AddByteArrays(elementsToAdd));
+   * @return the number of elements in the list
+   **/
+  public int llen() {
     return elementList.size();
   }
 
   /**
    * @param region the region this instance is stored in
-   * @param key the name of the set to add to
+   * @param key the name of the list to add to
    * @return the element actually popped
    */
   public byte[] lpop(Region<RedisKey, RedisData> region, RedisKey key) {
-    byte[] popped = elementRemove(0);
-    RemoveElementsByIndex removed = new RemoveElementsByIndex();
-    removed.add(0);
+    byte newVersion;
+    byte[] popped;
+    RemoveElementsByIndex removed;
+    synchronized (this) {
+      newVersion = incrementAndGetVersion();
+      popped = removeFirstElement();
+      removed = new RemoveElementsByIndex(newVersion);
```

Comment:
I think the sync can end right after the removeFirstElement call. No need to hold sync while creating the DeltaInfo instance